### ISSUE TITLE
[extension types] type-annotation-driven resolution?

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/language#3373]
The mental model that I have of type annotations in Dart is that they are only (_mostly?_) a tool for helping the type-checker type-check in cases where it doesn't have enough information to infer the types by itself, and would have to "fail" otherwise.

If a program A doesn't compile successfully, because it doesn't type-check, type annotations help with making A compile successfully.

Extension types appear to make it so that we can have a program B that compiles successfully, but we can use type annotations to change the runtime behavior of B. That is, changing the type annotation of a variable that is assigned a value that holds an extension type can change the meaning of a program (_not just in ways that makes it succeed or fail compilation_): 

```dart
void main() {
  final Foo i = Foo(2);
  print(i.toDouble()); // 0
  final int j = i; // We change the type annotation of i to int.
  print(j.toDouble()); // 2
  final k = i; // We remove the type annotation of i.
  print(k.toDouble()); // 0
}

extension type Foo(int i) implements int {
  double toDouble() {
    return 0.0;
  }
}
```

This is **very** surprising to me, because I am consciously relying on this not to be the case.

I'm used to being able to freely add/remove type annotations because of the assumption that this is an invariant program transformation with respect to the semantics of my program. (_something like, if it type-checks, it works, and adding more type annotations can help with making things work, but not change how things work._)

However, with extension types, this property will seemingly be lost. And any added/deleted type annotation would have to be carefully inspected to ensure that it doesn't change the meaning of my program beyond just a binary it compiles/it doesn't.

---

This also introduces confusing situations like the following:

Notice how `print(i.toDouble()` outputs a `0`, but if we print via `Bar`, we output a `2`, same with `k`:

```dart
void main() {
  final Foo i = Foo(2);
  print(i.toDouble()); // 0
  final int j = i;
  print(j.toDouble()); // 2
  final k = i;
  print(k.toDouble()); // 0
  Bar(i).output(); // contains i but outputs a 2
  Bar(j).output(); // contains j but outputs a 2
  Bar(k).output(); // contains k but output a 2 
}

extension type Foo(int i) implements int {
  double toDouble() {
    return 0.0;
  }
}

class Bar {
  final int i;
  
  Bar(this.i);
  
  void output() {
    print(i.toDouble());
  }
}
```

Of course, this probably is not surprising to the language team, but I wanted to point out that this is surprising to me as a user of Dart, and I think that it would be very difficult to explain this behavior to somebody that is new to programming or even just new to Dart.

---

Erik pointed out an important use-case for the ability to implement supertypes here (https://github.com/dart-lang/language/issues/3365#issuecomment-1735582023). While I do see the value in that, I've shared an opinion here (https://github.com/dart-lang/language/issues/3365#issuecomment-1735640181) that I don't think it would sufficiently solve the use-case of migrating from an old interface to a new one because a level of indirection would negate the safety guarantees that the current design would make possible.

To summarize, I don't think that being able to implement supertypes of the representation type "carries its weight" if it introduces such, in my opinion, "confusing" behaviors.

**Comment by jakemac53** [comment]
Regular extensions already introduce the same behavior difference you are describing - although it is a bit harder to construct an example:

```dart
void main() {
  int x = 1;
  print(x.a); // 0.0
  num y = x; // y gets the explicit type of `num`
  print(y.a); // 1.0
  var z = x; // z gets the inferred type of `int`
  print(z.a); // 0.0
}

extension _ on num {
  double get a => 1.0;
}

extension __ on int {
  double get a => 0.0;
}
```

**Comment by lrhn** [comment]
"Type annotations" define the declared type of variables or return types of functions.
Calling them "annotations" is really a holdover from Dart 1, where they could be ignored completely. Today they are significant, and affect a number of things, including which types you can promote to.

The declared type also defines which interface the value exposes. For statically resolved methods, like extension and extension type methods, it defines precisely which member is invoked.

The behavior you're describing is exactly the behavior that is intended, even desired.
Types matter, they can change the behavior of a program. As Jake points out, that's not new.

For example, I'll likely make myself a `FancyFuture` with a better typed error handling, which shadows `catchError`.
Because I want a different method, but without completely rewriting everything.

**Comment by mateusfccp** [comment]
As a Dart user, the current behavior is exactly what I would expect it to be.

I would never expect `final x = y;` to infer `x` as `int` instead of `Foo`.

**Comment by eernstg** [comment]
> type annotations in Dart .. are .. _(mostly?)_ .. a tool for helping the type-checker type-check in cases where it doesn't have enough information to infer the types by itself, and would have to "fail" otherwise.

I'd say that the primary role played by type annotations is to specify a contract between the provider and the clients of any given declaration, and getting machine-checked feedback on the correctness of the claimed properties.

Type inference can be used to reduce the verbosity of the code in cases where suitable types can be inferred.

Nominal types are associated with conceptual and application domain specific associations, and this implies that it would never be possible for a mechanical process to select suitable types in every case. For example, a compiler can't know whether an object that has two 64 bit long storage slots is intended to be an instance of `class Point` containing two `int` instance variables, or `class BinaryTreeNode` containing two object references, or a record of type `(double, bool)`, or any one of a thousand other plausible types for that representation.

So we must provide some amount of type information in our software, and then we can enjoy the convenience of having this information _propagated_ into some program constructs where type information has been omitted.

> Extension types appear to make it so that we can have a program B that compiles successfully, but we can use type annotations to change the runtime behavior of B.

Plain extensions have already been mentioned as a counterexample: This is not new. But there is more:

```dart
// Type arguments are reified, can be inferred, and they do make a difference.
void example1() {
  List<int> xs = [1]; // The behavior changes with the type annotation `List<num>`.
  dynamic d = xs; // Just so we know it's not just the static type.
  if (d is List<int>) print('Hello world!');
}

// Coercions.
void example2() {
  double n = 1; // Would be an `int` with `var n = 1;`, yielding different behavior.
  int Function(int) f = <X>(X x) => x; // Would be a generic function with `var f = ...`.
  int Function(int) g = Callable(); // Means `Callable().call` because of the type.
}

class Callable { int call(int i) => i + 1; }

// Syntax disambiguation.
void example3() {
  Set<Never> s = {}; // Would be a map with `var s = {};`.
}
```

That said, I'm actually quite worried about exactly this topic.

It would be great if we could clarify the distinction between statically resolved and dynamically resolved invocations. We might _visualize_ things like method invocations differently, such that the statically resolved invocations would have a light blue background in IDEs, and the dynamically resolved invocations could have a light green background (and dynamic invocations could have a bright red background ;-).

The green stuff would then be normal, statically checked OO code; the blue stuff would be invocations of top-level functions, extension methods, extension type methods, and other statically resolved constructs; and the red stuff would be unsafe code, using the type `dynamic` in various ways.

We'd then know that the static typing will fully determine the meaning of the blue stuff, whereas the green stuff has the same meaning under many different static types, and static types are basically ignored for the red stuff.

**Comment by lrhn** [comment]
Also, to quote the original code:
```dart
final int j = i; // We change the type annotation of i to int.
  ...
  final k = i; // We remove the type annotation of i.
```

Both of those comments are misleading, they sound as if the assignment changes the `i` variable.
They do not, instead they create new unrelated variables with the same value as the `i` variable currently has.
One of them has a different explicitly declared type, the other infers its declared type from the static type of `i`.

The `final k = i;` does not remove any type annotation, it just has an implicit type annotation instead of an explicit one. In the end, that makes no difference at all.

**Comment by modulovalue** [comment]
@lrhn said:

> Also, to quote the original code: [...]

I apologize, I did not mean to mislead, and I agree that I could have worded those comments in a more precise way. By "type annotation" I was referring to the syntactical "type labels" that users provide, not the type annotations that a fully type inferred AST would contain.

---

@lrhn said:

> For example, I'll likely make myself a FancyFuture with a better typed error handling, which shadows catchError.
Because I want a different method, but without completely rewriting everything.


For lack of a better expression, wouldn't this be like putting lipstick on a pig? I don't quite see the value of wrapping a `Future` in a `FancyFuture` just so that a suboptimal interface can look like an optimal one, but become a normal `Future` as soon as it is passed somewhere where a `Future` is expected.

I suppose you have considered providing a better `catchError` via a static extension. Why would you prefer wrapping a Future in a `FancyFuture` over search-and-replacing all `catchError` invocations with a `safeCatchError`?

Personally, I prefer increasing type-safety for some `dart:core` interfaces with extensions such as:

```dart
extension TypeSafeSet<K> on Set<K> {
  bool typed_contains(
    final K key,
  ) {
    return this.contains(key);
  }

  // ...
}
```

and this is working really well for me. I don't think I'd prefer wrapping all Sets in a `TypedSet` just to have a type-safer `contains` as this seems more involved than just using static extensions. Could I be missing something here?

**Comment by modulovalue** [comment]
Jake and Erik have provided examples that show that the extension type feature does not introduce  the type of behavior that I was referring to, and I agree with Jake and Erik, it doesn't.

But it would make it more widespread, and I do still think that this type of behavior is confusing _(in the sense that, I think, users would have little preexisting intuition for it that they could rely on for helping them understand it more easily)_.

As I've pointed out about the `FancyFuture` use-case [here](https://github.com/dart-lang/language/issues/3373#issuecomment-1742290636), and in a comment [here](https://github.com/dart-lang/language/issues/3365#issuecomment-1735640181) about a similar use-case, I don't think that supporting them provides enough value for it to be worth forcing users to learn the difference between statically resolved and dynamically resolved invocations. These concepts already exist in Dart, but Dart users can be highly productive without ever needing to understand them.

And even if an IDE could help with discovering which invocations are using which style (as Erik was suggesting), I think that it would be better not to force users to think about such details, and limiting confusing behavior to a minimum.

---

To summarize my argument:

- I think that being able to implement supertypes of the representation type in an extension type is very useful.
- but I think that supporting overrides of supertype members that are only overridden statically is not useful enough for it to introduce the confusing behaviors that I was referring to in the issue description.
